### PR TITLE
Feature/62 cheating

### PR DESCRIPTION
**Pull request by sosaidi**
_I have made the following updates:_
1. Implemented the buyCard function: The buyCard function now takes a Message object as an argument and extracts the cardId and cheatActivated values from it. It uses these values to buy a card from the market. If cheatActivated is true and the phone is turned over, the player gets the card for free. Otherwise, the player must pay the card's cost in turn coins.

2. Added the isPhoneTurnedOver method: This method uses the phone's accelerometer sensor to determine if the phone is turned over. It checks the phone's orientation and sets a flag accordingly.

3. Implemented the isTurnedOver method: This method registers a listener for the accelerometer sensor and sets a flag if the phone is turned over. It calculates the magnitude of the difference between the gravity and the accelerometer sensor's values to determine the orientation.

### **Explanation**
- The buyCard function checks if the player has enough turnCoins to buy the card and handles different scenarios based on cheat activation and phone orientation.
- The isPhoneTurnedOver method utilizes the phone's accelerometer sensor to determine the phone's orientation.
- The isTurnedOver method registers a listener for the accelerometer sensor and sets a flag based on the phone's orientation.
- The run method is not directly related to the buyCard or isPhoneTurnedOver methods, but it runs in a separate thread and checks if the phone is turned over.

### **Possible bugs:**
- NullPointer or NumberFormatException if the Message object doesn't contain the required keys.
- Potential issues with duplicate cards in the market's forPurchase list.
- IllegalStateException if the isPhoneTurnedOver method is called before setting the context.
- NullPointerException if the phone's accelerometer sensor is unavailable.
- Performance issues if the run method is not interrupted.

### **Possible improvements:**
- Input validation to ensure the Message object is valid.
- Using a Map object instead of CheatWrapper to store the cheat flag.
- Utilizing CountDownLatch instead of Thread.sleep in the isTurnedOver method.
- Adding a timeout mechanism to prevent the isTurnedOver method from running indefinitely.

### CODE REVIEW

**Review by Benjamin-Dippold — CHANGES_REQUESTED** [review]
I am having difficulties testing this feature. When trying to host a server (specifically after pressing "Show my IP Address") the application crashes. This behaviour was observed on multiple devices.
image
Is this something that has come up in development, or is this a device specific problem?

Since I cannot try out the new feature and test it's functionality, I am unable to review it's changes.

File: app/src/test/java/at/vunfer/openrealms/model/PlayAreaTest.java [hunk 1]
```diff
@@ -285,6 +284,29 @@ void testBuyCard() {
         assertTrue(player1.getPlayArea().getPlayerCards().getDiscardedCards().contains(toBuy));
     }
 
+    /*@Test
```
Comment (Kranosos) [inline app/src/test/java/at/vunfer/openrealms/model/PlayAreaTest.java:287]: Should this commented Test be removed?

Comment (Kranosos) [inline app/src/test/java/at/vunfer/openrealms/model/PlayAreaTest.java:287]: Should this commented out test be removed

File: app/src/main/java/at/vunfer/openrealms/network/DataKey.java [hunk 2]
```diff
@@ -8,6 +8,8 @@ public enum DataKey {
     CHOICE,
     PLAYER_STATS, // used to specify the player stats of a player
     OPTIONS,
+    CHEAT_ACTIVATE,
+    ERROR_MESSAGE,
```
Comment (Kranosos) [inline app/src/main/java/at/vunfer/openrealms/network/DataKey.java:12]: What is this DataKey used for? is it used at all?

**Review by Kranosos — APPROVED** [review]
I really like the subtility of the cheating feature and think this will prove to be a very good addition to our game and to the player experience. The quality gates seem to look fine, although they could use a little more coverage!
Overall…LGTM!

**Comment by Ne0Strix** [comment]
Cheating now works by shaking the device. Once it's enabled, the ring of the coin-turnvalues turns white on both devices. The opponent can then tap that ring to uncover the attempt. If that happens before the turn is ended, all cards bought using the cheating feature are sacrificed (removed from the game) and a preset amount of damage (10) is dealt the cheating player. If the opponent wrongly accuses the player of cheating they're dealt another preset amount of damage (5).

**Review by Benjamin-Dippold — APPROVED** [review]
LGTM

**Pull request merged**